ntoskrnl/include/internal/i386/mm.h

```diff
@@ -3,43 +3,38 @@
  */
 #pragma once
 
+#define _MI_PAGING_LEVELS 2
```

Comment:
I don't understand why PAE should not use `_MI_PAGING_LEVELS == 3`. Can you elaborate?
Setting it to 2 for all i386 means we will need to write code _specifically_ for PAE, which we absolutely want to avoid if possible

Reply:
https://jira.reactos.org/browse/CORE-16702?focusedCommentId=126164&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-126164

"Paging (or linear-address translation) is the process of translating linear addresses so that they can be used to access memory or I/O devices. Paging translates each linear address to a physical addressand determines, for each translation, what accesses to the linear address are allowed (the address’s access rights) and the type of caching used for such 
accesses (the address’s memory type)." - Intel® 64 and IA-32 Architectures Software Developer’s Manual Volume 3: System Programming Guide. Chapter 4 PAGING.

According to the specification, there are three paging modes (can also say that there is a fourth mode - this is when paging is not used):
1) 32-bit paging
2) PAE paging
3) IA-32e paging - available only on processors that support the Intel 64 architecture.

PAE requires (in addition to 32-bit paging mode) a 32-byte table with four entries (pointed to by CR3). Logical processor maintains a set of four (4) internal, non-architectural registers, called PDPTE0, PDPTE1, PDPTE2, and PDPTE3. The logical processor loads these registers from the 4 elements of this table in memory.
Therefore PAE paging mode does not require a third level of tables (for MM context). This is only required for IA-32e paging (64-bit).

Reply:
My opinion is to first make (MM + CC) functionality for 32-bit mode, but it's too early to talk about PAE.

Reply:
> My opinion is to first make (MM + CC) functionality for 32-bit mode, but it's too early to talk about PAE.

You're proposing this change here, that's why I'm asking. We can move the discussion into the next (possible) pull request with full PAE implementation, but in that case please revert the change :)

In the ticket you wrote:
> In the context of a memory manager (MM), PAE also uses two levels of tables. The difference is that there is additionally a small (4 elements) table for loading CR3. But for MM it is transparent and not used.

Do you want to say that MM is going to load a constant values into PD table and won't change them while the system is running? Is it how Windows works?

I, personally, don't see much difference with Long mode (IA-32e) paging here - we just have a 4-entry PD table instead of a 512-entry table. Where am I making a mistake?

Reply:
1) No problems :)

2) The processor has four internal registers. They contain the physical addresses of the PDs from the 4-entry table. It is not MM who does this (most likely KE's job). Logically, this should happen with every process change. But this isn't MM. This is the difference. In IA-32e, no internal registers are used and tables are maintained by MM.

Generally PAE mode is an add-on over 32-bit paging mode. The difference is that in PAE mode, you can use a larger number of pages of physical memory (that's what everything was intended for). But both modes use 32 - bit virtual address space. This can be called a "clever move" by Intel engineers :). If you can outsmart them :), then you can make universal code. But I would not spend a lot of time on this (at least at this stage). I probably agree with Timo that x64 is now more important than PAE (but at the same time, up to x64 (in priority) are HAL, MM, CC, PNP, PO).

If you are serious about PAE, then you need to start with the OS bootloader (FreeLoader).

Reply:
> But I would not spend a lot of time on this (at least at this stage). I probably agree with Timo that x64 is now more important than PAE

Than please revert the change. We will continue the discussion some time in future